contracts/YieldStrategyLever.sol

```diff
@@ -0,0 +1,554 @@
+
+        // Give the vault to the contract
+        giver.seize(vaultId, address(this));
+
+        IPool pool = IPool(LADLE.pools(seriesId));
+
+        // Check if we're pre or post maturity.
+        bool success;
+        if (uint32(block.timestamp) < CAULDRON.series(seriesId).maturity) {
+            IFYToken fyToken = pool.fyToken();
+            // Repay:
+            // Series is not past maturity.
+            // Borrow to repay debt, move directly to the pool.
+            bytes memory data = bytes.concat(
+                bytes1(bytes1(uint8(uint256(Operation.REPAY)))), // [0:1]
+                seriesId, // [1:7]
+                vaultId, // [7:19]
+                ilkId, // [19:25]
+                bytes32(ink), // [25:57]
+                bytes32(art) // [57:89]
+            );
+            success = IERC3156FlashLender(address(fyToken)).flashLoan(
+                this, // Loan Receiver
+                address(fyToken), // Loan Token
+                art, // Loan Amount: borrow exactly the debt to repay.
+                data
+            );
+        } else {
+            FlashJoin join = FlashJoin(
+                address(LADLE.joins(seriesId & ASSET_ID_MASK))
+            );
+            IERC20 baseAsset = IERC20(pool.base());
+            uint256 depositIntoJoin = baseAsset.balanceOf(address(join)) -
+                join.storedBalance();
+
+            // Close:
+            // Series is past maturity, borrow and move directly to collateral pool.
+            bytes memory data = bytes.concat(
+                bytes1(bytes1(uint8(uint256(Operation.CLOSE)))), // [0:1]
+                seriesId, // [1:7]
+                vaultId, // [7:19]
+                ilkId, // [19:25]
+                bytes32(ink), // [25:57]
+                bytes32(art) // [57:89]
+            );
+            // We have a debt in terms of fyWEth, but should pay back in WEth.
+            // `base` is how much WEth we should pay back.
+            uint128 base = CAULDRON.debtToBase(seriesId, art.u128());
+            success = join.flashLoan(
+                this, // Loan Receiver
+                address(baseAsset), // Loan Token
+                base, // Loan Amount
+                data
+            );
+
+            // At this point, we have only Weth left. Hopefully: this comes
+            // from the collateral in our vault!
+
+            // There is however one caveat. If there was Weth in the join to
+            // begin with, this will be billed first. Since we want to return
+            // the join to the starting state, we should deposit tokens back.
+            // The amount is simply what was in it before, minus what is still
+            // in it. The calculation is as `available` in the Join contract.
+            depositIntoJoin +=
+                join.storedBalance() -
+                baseAsset.balanceOf(address(join));
+            baseAsset.safeTransfer(address(join), depositIntoJoin);
```

Comment:
I don't get this bit. `join.storedBalance()` can't be more than `baseAsset.balanceOf(address(join))`.

Also, if there were any assets in the join that were unaccounted for, it's ok for the caller to get them as a bonus.

Reply:
In earlier iteration we had decided that we would leave joins in the same state that we found them in that's why the check.